subsys/net/lib/sockets/socketpair.c

```diff
@@ -0,0 +1,1132 @@
+};
+
+#define SPAIR_FLAGS_DEFAULT 0
+
+/**
+ * Socketpair endpoint structure
+ *
+ * This structure represents one half of a socketpair (an 'endpoint').
+ *
+ * The implementation strives for compatibility with socketpair(2).
+ *
+ * Resources contained within this structure are said to be 'local', while
+ * reources contained within the other half of the socketpair (or other
+ * endpoint) are said to be 'remote'.
+ *
+ * Theory of operation:
+ * - each end of a socketpair owns a @a recv_q
+ * - since there is no write queue, data is either written or not
+ * - read and write operations may return partial transfers
+ * - read operations may block if the local @a recv_q is empty
+ * - write operations may block if the remote @a recv_q is full
+ * - each endpoint may be blocking or non-blocking
+ */
+struct spair {
+	int remote; /**< the remote endpoint file descriptor */
+	u32_t flags; /**< status and option bits */
+	struct k_sem sem; /**< semaphore for exclusive structure access */
+	struct k_pipe recv_q; /**< receive queue of local endpoint */
+	/** indicates write of local @a recv_q occurred */
+	struct k_poll_signal write_signal;
+	/** indicates read of local @a recv_q occurred */
+	struct k_poll_signal read_signal;
+	/** buffer for @a recv_q recv_q */
+	u8_t buf[CONFIG_NET_SOCKETPAIR_BUFFER_SIZE];
+};
+
+/* forward declaration */
+static const struct socket_op_vtable spair_fd_op_vtable;
+
+#undef sock_is_nonblock
+/** Determine if a @ref spair is in non-blocking mode */
+static inline bool sock_is_nonblock(const struct spair *spair)
+{
+	return !!(spair->flags & SPAIR_FLAG_NONBLOCK);
+}
+
+/** Determine if a @ref spair is connected */
+static inline bool sock_is_connected(const struct spair *spair)
+{
+	const struct spair *remote = z_get_fd_obj(spair->remote,
+		(const struct fd_op_vtable *)&spair_fd_op_vtable, 0);
+
+	if (remote == NULL) {
+		return false;
+	}
+
+	return true;
+}
+
+#undef sock_is_eof
+/** Determine if a @ref spair has encountered end-of-file */
+static inline bool sock_is_eof(const struct spair *spair)
+{
+	return !sock_is_connected(spair);
+}
+
+/**
+ * Determine bytes available to write
+ *
+ * Specifically, this function calculates the number of bytes that may be
+ * written to a given @ref spair without blocking.
+ */
+static inline size_t spair_write_avail(struct spair *spair)
+{
+	struct spair *const remote = z_get_fd_obj(spair->remote,
+		(const struct fd_op_vtable *)&spair_fd_op_vtable, 0);
+
+	struct k_pipe *const p = (remote == NULL)
+		? NULL
+		: &remote->recv_q;
+
+	if (remote == NULL) {
+		return 0;
+	}
+
+	return k_pipe_write_avail(p);
+}
+
+/**
+ * Determine bytes available to read
+ *
+ * Specifically, this function calculates the number of bytes that may be
+ * read from a given @ref spair without blocking.
+ */
+static inline size_t spair_read_avail(struct spair *spair)
+{
+	struct k_pipe *const p = &spair->recv_q;
+
+	return k_pipe_read_avail(p);
```

Comment:
do we need this `p` variable here, why not say
```
return k_pipe_read_avail(&spair->recv_q);
```

Reply:
Same.